File: src/main/java/core/procedures/math/trigonometry/Cos.java
package core.procedures.math.trigonometry;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.procedures.math.Multiplication;
import core.scm.BigComplex;
import core.scm.BigRatio;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class Cos extends AFn {

  public Cos() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "cos";
  }

  @Override
  public Number apply1(Object arg) {
    return cos((Number)arg);
  }

  public static Number cos(Number number) {
    /* Special cases */
    if (Utils.INSTANCE.isZero(number)) {
      return 1L;
    }
    if (number instanceof BigDecimal) {
      return cos((BigDecimal)number);
    } else if (number instanceof BigInteger) {
      return cos((BigInteger)number);
    } else if (number instanceof BigComplex) {
      return Cos.cos((BigComplex)number);
    } else if (number instanceof BigRatio) {
      return cos(((BigRatio)number).toBigDecimal());
    }
    return Math.cos(number.doubleValue());
  }

  public static double cos(BigDecimal bd) {
    double v = bd.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.cos(v);
    }
  }

  public static double cos(BigInteger bi) {
    double v = bi.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.cos(v);
    }
  }

  public static BigComplex cos(BigComplex c) {
    BigDecimal re = c.getRe();
    BigDecimal im = c.getIm();
    return new BigComplex(Multiplication.Companion.apply(Cos.cos(re), Cosh.cosh(im)),
                          Multiplication.Companion.apply(-1d, Multiplication.Companion.apply(Sin.sin(re), Sinh.sinh(im))));
  }
}

File: src/main/java/core/procedures/math/trigonometry/Cos.kt
package core.procedures.math.trigonometry

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.procedures.math.Multiplication
import core.scm.Complex
import core.utils.Utils
import kotlin.math.cos
import kotlin.math.cosh
import kotlin.math.sin
import kotlin.math.sinh

class Cos : AFn<Number?, Number>(name = "cos", isPure = true, arity = Exactly(1),
                                 mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?) = when {
        Utils.isZero(arg!!) -> 1L
        arg is Complex   -> cos(arg)
        else                -> cos(arg.toDouble())
    }

    companion object {

        private val multiplication = Multiplication()

        fun cos(c: Complex): Complex {
            val x = c.re.toDouble()
            val y = c.im.toDouble()
            return Complex(multiplication.invoke(cos(x), cosh(y)), multiplication.invoke(-1.0, multiplication.invoke(sin(x), sinh(y))))
        }
    }
}
